xchange-stream-binance/src/main/java/info/bitrich/xchangestream/binance/BinanceStreamingMarketDataService.java

```diff
@@ -300,41 +325,204 @@ private void initRawOrderBookUpdatesSubscription(CurrencyPair currencyPair) {
+
+    /**
+     * Helps to keep integrity of book snapshot which is initialised and patched on different
+     * threads.
+     */
+    private final Object monitor = new Object();
+
+    private OrderBookSubscription(
+        Observable<DepthBinanceWebSocketTransaction> deltasObservable, CurrencyPair currencyPair) {
+      this.deltasObservable = deltasObservable;
+      this.currencyPair = currencyPair;
+    }
 
-    private OrderbookSubscription(Observable<DepthBinanceWebSocketTransaction> stream) {
-      this.stream = stream;
+    public Observable<OrderBook> connect() {
+      if (isDisposed())
+        throw new IllegalStateException(
+            "Disposed before, use a new instance to connect next time.");
+
+      disposables.add(asyncInitialiseOrderBookSnapshot());
+
+      deltasObservable
+          .doOnNext(
+              delta -> {
+                synchronized (monitor) {
+                  if (isBookInitialised()) {
+                    if (!appendDelta(delta)) {
+                      disposables.add(asyncInitialiseOrderBookSnapshot());
+                    }
+                  } else {
+                    bufferDelta(delta);
+                  }
+                }
+              })
+          .filter(delta -> isBookInitialised())
+          .map(delta -> getBook())
+          .doFinally(() -> dispose())
+          .subscribe(booksSubject);
+
+      return booksSubject.hide();
     }
 
-    void invalidateSnapshot() {
-      snapshotLastUpdateId.set(0);
+    @Override
+    public void dispose() {
+      if (!isDisposed()) {
+        booksSubject.onComplete();
+        disposables.dispose();
+      }
     }
 
-    void initSnapshotIfInvalid(CurrencyPair currencyPair) {
-      if (snapshotLastUpdateId.get() != 0) return;
-      try {
-        LOG.info("Fetching initial orderbook snapshot for {} ", currencyPair);
-        onApiCall.run();
-        fallbackOnApiCall.get().run();
-        BinanceOrderbook book = fetchBinanceOrderBook(currencyPair);
-        snapshotLastUpdateId.set(book.lastUpdateId);
-        lastUpdateId.set(book.lastUpdateId);
-        orderBook = BinanceMarketDataService.convertOrderBook(book, currencyPair);
-      } catch (Exception e) {
-        LOG.error("Failed to fetch initial order book for " + currencyPair, e);
-        snapshotLastUpdateId.set(0);
-        lastUpdateId.set(0);
-        orderBook = null;
+    private void disposeWithError(Throwable error) {
+      if (!isDisposed()) {
+        booksSubject.onError(error);
+        disposables.dispose();
+      }
+    }
+
+    @Override
+    public boolean isDisposed() {
+      return booksSubject.hasComplete() || booksSubject.hasThrowable();
+    }
+
+    private boolean isBookInitialised() {
+      return getBook() != null;
+    }
+
+    private OrderBook getBook() {
+      return bookRef.get();
+    }
+
+    private void bufferDelta(DepthBinanceWebSocketTransaction delta) {
+      synchronized (monitor) {
```

Comment:
why are you synchronizing here if you already use a `ConcurrentLinkedQueue`

Reply:
To explain my thinking:
- atomic references and concurrent queue ensure thread safety on level of these particular objects (in isolation, every object is thread safe for itself)
- in our case the entire state of `OrderBookSubscription` needs to remain in sync
- there is some logic which involves multiple of these individualy thread safe objects, thus entire conditions need to be atomic, they are wrapped with `synchronized (monitor) { ...}`
- I did it more by convention:
-- Logic on multiple objects? Synchronise that with the monitor
-- Accessing one of these objects like current snapshot of an order book? Then skip `synchronisation` of the entire state of the `OrderBookSubscription`, and rely just on thread safety provided by the atomic reference:
```
private OrderBook getBook() {
   return bookRef.get();
}
```

@earce At the end. It's true, I could remove some of synchronisation, I think it would be a bit less schematic, and more prone to bugs caused by further maintainers, but its so small risk, I can modify the code as you prefer.

Reply:
Update: I have simplified thread synchronisation, more in other comments.